README.md
<p align="center">
    <a href="https://github.com/yii2tech" target="_blank">
        <img src="https://avatars2.githubusercontent.com/u/12951949" height="100px">
    </a>
    <h1 align="center">Site Map Extension for Yii 2</h1>
    <br>
</p>

This extension provides support for site map and site map index files generating.

For license information check the [LICENSE](LICENSE.md)-file.

[![Latest Stable Version](https://img.shields.io/packagist/v/yii2tech/sitemap.svg)](https://packagist.org/packages/yii2tech/sitemap)
[![Total Downloads](https://img.shields.io/packagist/dt/yii2tech/sitemap.svg)](https://packagist.org/packages/yii2tech/sitemap)
[![Build Status](https://travis-ci.org/yii2tech/sitemap.svg?branch=master)](https://travis-ci.org/yii2tech/sitemap)


Installation
------------

The preferred way to install this extension is through [composer](http://getcomposer.org/download/).

Either run

```
php composer.phar require --prefer-dist yii2tech/sitemap
```

or add

```json
"yii2tech/sitemap": "*"
```

to the require section of your composer.json.


Usage
-----

This extension provides support for site map and site map index files generation.
You can use `\yii2tech\sitemap\File` for site map file composition:

```php
<?php

use yii2tech\sitemap\File;

$siteMapFile = new File();

$siteMapFile->writeUrl(['site/index'], ['priority' => '0.9']);
$siteMapFile->writeUrl(['site/about'], ['priority' => '0.8', 'changeFrequency' => File::CHECK_FREQUENCY_WEEKLY]);
$siteMapFile->writeUrl(['site/signup'], ['priority' => '0.7', 'lastModified' => '2015-05-07']);
$siteMapFile->writeUrl(['site/contact']);

$siteMapFile->close();
```

In case you put sitemap generation into a console command, you will need to manually configure URL manager
parameters for it. For example:

```php
<?php

return [
    'id' => 'my-console-application',
    'components' => [
        'urlManager' => [
            'hostInfo' => 'https://example.com',
            'baseUrl' => '/',
            'scriptUrl' => '/index.php',
        ],
        // ...
    ],
    // ...
];
```


## Creating site map index files <span id="creating-site-map-index-files"></span>

There is a limitation on the site map maximum size. Such file can not contain more then 50000 entries and its
actual size can not exceed 50MB. If you web application has more then 50000 pages and you need to generate
site map for it, you'll have to split it between several files and then generate a site map index file.
It is up to you how you split your URLs between different site map files, however you can use `\yii2tech\sitemap\File::getEntriesCount()`
or `\yii2tech\sitemap\File::getIsEntriesLimitReached()` method to check count of already written entries.

For example: assume we have an 'item' table, which holds several millions of records, each of which has a detail
view page at the web application. In this case generating site map files for such pages may look like following:

```php
<?php

use app\models\Item;
use yii2tech\sitemap\File;

$query = Item::find()->select(['slug'])->asArray();

$siteMapFileCount = 0;
foreach ($query->each() as $row) {
    if (empty($siteMapFile)) {
        // if there is no active file - create one with unique name:
        $siteMapFile = new File();
        $siteMapFileCount++;
        $siteMapFile->fileName = 'item_' . $siteMapFileCount . '.xml';
    }

    $siteMapFile->writeUrl(['item/view', 'slug' => $row['slug']]);
    if ($siteMapFile->getIsEntriesLimitReached()) {
        // once file is full - close it, allowing creating a new one at the next cycle iteration:
        unset($siteMapFile);
    }
}
```

Once all site map files are generated, you can compose index file, using the following code:

```php
<?php

use yii2tech\sitemap\IndexFile;

$siteMapIndexFile = new IndexFile();
$siteMapIndexFile->writeUp();
```

> Note: by default site map files are stored under the path '@app/web/sitemap'. If you need a different file path
  you should adjust `fileBasePath` field accordingly.


## Rendering on-the-fly <span id="rendering-on-the-fly"></span>

Saving sitemap to the physical file may be not a best option to keep it up-to-date. Such file should be manually re-created,
once some changes among site pages appear. You may setup a web controller, which will render 'sitemap.xml' file on demand,
once it is been requested. This controller may apply caching and its busting logic.
First of all, you'll have to set up a route for the controller action rendering the sitemap in your URL manager. For example:

```php
<?php

return [
    'components' => [
        'urlManager' => [
            'rules' => [
                'sitemap.xml' => 'site/sitemap',
                // ...
            ],
        ],
        // ...
    ],
    // ...
];
```

Then you'll need to create an action, which will render sitemap file content and emit it to the web client.
You can use PHP 'in memory' stream as a file name for the sitemap file during its composition.
The final implementation may look like following:

```php
<?php

namespace app\controllers;

use Yii;
use yii\web\Controller;
use yii\web\Response;
use yii2tech\sitemap\File;

class SiteController extends Controller
{
    public function actionSitemap()
    {
        // get content from cache:
        $content = Yii::$app->cache->get('sitemap.xml');
        if ($content === false) {
            // if no cached value exists - create an new one
            // create sitemap file in memory:
            $sitemap = new File();
            $sitemap->fileName = 'php://memory';
            
            // write your site URLs:
            $sitemap->writeUrl(['site/index'], ['priority' => '0.9']);
            // ...
            
            // get generated content:
            $content = $sitemap->getContent();

            // save generated content to cache
            Yii::$app->cache->set('sitemap.xml', $content);
        }

        // send sitemap content to the user agent:
        $response = Yii::$app->getResponse();
        $response->format = Response::FORMAT_RAW;
        $response->getHeaders()->add('Content-Type', 'application/xml;');
        $response->content = $content;
        
        return $response;
    }
}
```


## Customizing file envelope <span id="customizing-file-envelope"></span>

You can customize entries envelope for the particular file using following options:

 - `\yii2tech\sitemap\BaseFile::$header` - content, which should be written at the beginning of the file, once it has been opened;
 
 - `\yii2tech\sitemap\BaseFile::$footer` - content, which should be written at the end of the file before it is closed;
 
 - `\yii2tech\sitemap\BaseFile::$rootTag` - defines XML root tag name and attributes;
 
For example:

```php
<?php

use yii2tech\sitemap\File;

$siteMapFile = new File([
    'header' => '<?xml version="1.0" encoding="UTF-8"?><?xml-stylesheet type="text/xsl" href="//example.com/main-sitemap.xsl"?>',
    'rootTag' => [
        'tag' => 'urlset',
        'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9',
        'xmlns:xsi' => 'http://www.w3.org/2001/XMLSchema-instance',
        'xmlns:image' => 'http://www.google.com/schemas/sitemap-image/1.1',
        'xsi:schemaLocation' => 'http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd http://www.google.com/schemas/sitemap-image/1.1 http://www.google.com/schemas/sitemap-image/1.1/sitemap-image.xsd',
    ],
]);

$siteMapFile->writeUrl(['site/index'], ['priority' => '0.9']);
// ...

$siteMapFile->close();
```


## Rendering non-standard tags <span id="rendering-non-standard-tags"></span>

While there is a [standard](http://www.sitemaps.org/), which defines sitemap content, particular search engines may accept
extra tags and options. The most widely used are image and video descriptions.
Method `\yii2tech\sitemap\File::writeUrl()` supports rendering image and video information.

For adding images to the sitemap entry use 'images' option. For example:

```php
<?php

use yii2tech\sitemap\File;

$siteMapFile = new File([
    'rootTag' => [
        'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9',
        'xmlns:image' => 'http://www.google.com/schemas/sitemap-image/1.1', // you will need to add XML namespace for non-standard tags
    ],
]);

$siteMapFile->writeUrl(['site/index'], [
    'images' => [
        [
            'url' => 'http://example.com/images/logo.jpg',
            'title' => 'Logo',
        ],
        [
            'url' => 'http://example.com/images/avatars/john-doe.jpg',
            'title' => 'Author',
        ],
        // ...
    ],
]);
// ...

$siteMapFile->close();
```

For adding videos to the sitemap entry use 'videos' option. For example:

```php
<?php

use yii2tech\sitemap\File;

$siteMapFile = new File([
    'rootTag' => [
        'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9',
        'xmlns:video' => 'http://www.google.com/schemas/sitemap-video/1.1', // you will need to add XML namespace for non-standard tags
    ],
]);

$siteMapFile->writeUrl(['site/index'], [
    'videos' => [
        [
            'title' => 'Demo video',
            'description' => 'Demo video of the main process',
            'thumbnailUrl' => 'http://example.com/images/demo-video.jpg',
            'player' => [
                'url' => 'http://example.com/videos/demo.flv',
                'allowEmbed' => true,
                'autoplay' => 'ap=1',
            ],
            'publicationDate' => '2019-08-02',
            'duration' => 240,
        ],
        [
            'title' => 'Our team',
            'description' => 'Greetings from our team',
            'thumbnailUrl' => 'http://example.com/images/our-team.jpg',
            'player' => [
                'url' => 'http://example.com/videos/our-team.flv',
                'allowEmbed' => true,
                'autoplay' => 'ap=1',
            ],
            'publicationDate' => '2019-08-02',
            'duration' => 120,
        ],
        // ...
    ],
]);
// ...

$siteMapFile->close();
```

You can also add any custom content to the URL tag using 3rd argument of the `\yii2tech\sitemap\File::writeUrl()` method.
For example:

```php
<?php

use yii2tech\sitemap\File;

$siteMapFile = new File([
    'rootTag' => [
        'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9',
        'xmlns:image' => 'http://www.google.com/schemas/sitemap-image/1.1', // you will need to add XML namespace for non-standard tags
    ],
]);

$siteMapFile->writeUrl(
    ['site/index'],
    [],
    '<image:image><image:loc>http://example.com/images/logo.jpg</image:loc></image:image>'
);
// ...

$siteMapFile->close();
```

**Heads up!** Remember that you'll have to add corresponding XML namespaces to the sitemap file, using `\yii2tech\sitemap\BaseFile::$rootTag`,
in order to non-standard tags being recognized by the search engines.


CHANGELOG.md
Yii 2 Site Map extension Change Log
===================================

1.1.1, October 17, 2019
-----------------------

- Bug #7: Fixed content returned by `File::getContent()` does not contain closing tag and footer (klimov-paul)


1.1.0, August 2, 2019
---------------------

- Enh: Removed `yii\base\Object::className()` in favor of native PHP syntax `::class`, which does not trigger autoloading (klimov-paul)
- Enh: Added support for 'images' and 'videos' options to `File::writeUrl()` (klimov-paul)
- Enh: Added ability to pass extra XML content to `File::writeUrl()` (klimov-paul)
- Enh: Extracted special `LimitReachedException` exception class (klimov-paul)
- Enh: Added ability to use PHP stream as `BaseFile::$fileName` (klimov-paul)
- Enh #5: Added `header`, `footer` and `rootTag` fields to `BaseFile` allowing customizing of the file entries envelope (GeniJaho, klimov-paul)
- Enh #6: Added `BaseFile::$lineBreak` allowing setup of the lines separator (easelify)


1.0.2, January 24, 2019
-----------------------

- Enh #3: Sitemap file name added to the message of the "max entries exceed" exception (machour, klimov-paul)
- Enh #4: Sitemap files now skip rendering of the optional tags if their value is `null` (OndrejVasicek, klimov-paul)


1.0.1, November 3, 2017
-----------------------

- Bug: Usage of deprecated `yii\base\Object` changed to `yii\base\BaseObject` allowing compatibility with PHP 7.2 (klimov-paul)


1.0.0, December 26, 2015
------------------------

- Initial release.


tests/FileTest.php
<?php

namespace yii2tech\tests\unit\sitemap;

use yii2tech\sitemap\File;
use yii2tech\sitemap\LimitReachedException;

/**
 * @see File
 */
class FileTest extends TestCase
{
    /**
     * Creates the site map file instance.
     * @return File
     */
    protected function createSiteMapFile()
    {
        $siteMapFile = new File();
        $siteMapFile->fileBasePath = $this->getTestFilePath();
        return $siteMapFile;
    }

    // Tests:

    public function testWriteBasicXml()
    {
        $siteMapFile = $this->createSiteMapFile();

        $siteMapFile->write('');
        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains('<?xml', $fileContent);
        $this->assertContains('<urlset', $fileContent);
        $this->assertContains('</urlset>', $fileContent);
    }

    public function testWriteUrl()
    {
        $siteMapFile = $this->createSiteMapFile();

        $testUrl = 'http://test.url';
        $testLastModifiedDate = '2010-07-15';
        $testChangeFrequency = 'test_frequency';
        $testPriority = 0.2;
        $siteMapFile->writeUrl($testUrl, [
            'lastModified' => $testLastModifiedDate,
            'changeFrequency' => $testChangeFrequency,
            'priority' => $testPriority
        ]);

        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains('<url', $fileContent);
        $this->assertContains('</url>', $fileContent);
        $this->assertContains($testUrl, $fileContent, 'XML does not contains the URL!');
        $this->assertContains($testLastModifiedDate, $fileContent, 'XML does not contains the last modified date!');
        $this->assertContains($testChangeFrequency, $fileContent, 'XML does not contains the change frequency!');
        $this->assertContains((string)$testPriority, $fileContent, 'XML does not contains the priority!');
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteUrlParams()
    {
        $siteMapFile = $this->createSiteMapFile();

        $siteMapFile->writeUrl(['controller/action']);

        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());
        $this->assertContains('http://test.com/index.php?r=controller%2Faction', $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteUrlDefaultOptions()
    {
        $siteMapFile = $this->createSiteMapFile();
        $siteMapFile->defaultOptions = [
            'lastModified' => '2010-01-01',
            'changeFrequency' => 'test_frequency',
            'priority' => '0.1'
        ];

        $siteMapFile->writeUrl('http://test.url');
        $siteMapFile->close();
        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains($siteMapFile->defaultOptions['lastModified'], $fileContent);
        $this->assertContains($siteMapFile->defaultOptions['changeFrequency'], $fileContent);
        $this->assertContains($siteMapFile->defaultOptions['priority'], $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testEntriesCountIncrement()
    {
        $siteMapFile = $this->createSiteMapFile();

        $originalEntriesCount = $siteMapFile->getEntriesCount();
        $this->assertEquals(0, $originalEntriesCount, 'Original entries count is wrong!');

        $testUrl = 'http://test.url';
        $siteMapFile->writeUrl($testUrl);

        $postWriteEntriesCount = $siteMapFile->getEntriesCount();

        $this->assertEquals($originalEntriesCount + 1, $postWriteEntriesCount, 'No entries count increment detected!');
        $siteMapFile->close();
    }

    /**
     * @depends testEntriesCountIncrement
     */
    public function testEntiesCountExceedException()
    {
        $siteMapFile = $this->createSiteMapFile();

        $this->expectException(LimitReachedException::class);
        for ($i = 1; $i < $siteMapFile->maxEntriesCount + 2; $i++) {
            $siteMapFile->writeUrl('http://test.url');
        }
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteUrlWithDefaultOptions()
    {
        $siteMapFile = $this->createSiteMapFile();

        $testUrl = 'http://test.url';
        $siteMapFile->writeUrl($testUrl);

        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains("<url><loc>{$testUrl}</loc></url>", $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteUrlWithInvalidOption()
    {
        $siteMapFile = $this->createSiteMapFile();

        $this->expectException(\yii\base\InvalidArgumentException::class);

        $siteMapFile->writeUrl('http://test.url', [
            'invalidOption' => 'some-value',
        ]);
    }

    /**
     * @depends testWriteBasicXml
     */
    public function testCustomizeEnvelope()
    {
        $siteMapFile = $this->createSiteMapFile();
        $siteMapFile->header = '<!-- header -->';
        $siteMapFile->footer = '<!-- footer -->';
        $siteMapFile->rootTag = [
            'tag' => 'myurlset',
            'xmlns' => 'http://example.com',
        ];

        $siteMapFile->write('');
        $siteMapFile->close();

        $fileContent = str_replace($siteMapFile->lineBreak, '', file_get_contents($siteMapFile->getFullFileName()));

        $expectedContent = '<!-- header --><myurlset xmlns="http://example.com"></myurlset><!-- footer -->';

        $this->assertSame($expectedContent, $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteImages()
    {
        $siteMapFile = $this->createSiteMapFile();

        $siteMapFile->writeUrl('http://example.com/some', [
            'images' => [
                [
                    'url' => 'http://example.com/images/1.jpg',
                    'title' => 'test title',
                    'caption' => 'test caption',
                    'geoLocation' => 'test location',
                    'license' => 'test license',
                ],
                [
                    'url' => 'http://example.com/images/2.jpg',
                ],
            ],
        ]);

        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains('<image:image>', $fileContent);
        $this->assertContains('</image:image>', $fileContent);

        $this->assertContains('<image:loc>http://example.com/images/1.jpg</image:loc>', $fileContent);
        $this->assertContains('<image:loc>http://example.com/images/2.jpg</image:loc>', $fileContent);

        $this->assertContains('<image:title>test title</image:title>', $fileContent);
        $this->assertContains('<image:caption>test caption</image:caption>', $fileContent);
        $this->assertContains('<image:geo_location>test location</image:geo_location>', $fileContent);
        $this->assertContains('<image:license>test license</image:license>', $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteVideos()
    {
        $siteMapFile = $this->createSiteMapFile();

        $siteMapFile->writeUrl('http://example.com/some', [
            'videos' => [
                [
                    'title' => 'test title',
                    'description' => 'test description',
                    'thumbnailUrl' => 'http://example.com/images/thumbnail.jpg',
                    'player' => [
                        'url' => 'http://example.com/videos/1.flv',
                        'allowEmbed' => true,
                        'autoplay' => 'ap=1',
                    ],
                    'publicationDate' => '2019-08-02',
                    'duration' => 120,
                ],
                [
                    'player' => [
                        'url' => 'http://example.com/videos/2.flv',
                        'allowEmbed' => true,
                        'autoplay' => 'ap=1',
                    ],
                ],
            ],
        ]);

        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains('<video:video>', $fileContent);
        $this->assertContains('</video:video>', $fileContent);

        $this->assertContains('<video:player_loc allow_embed="yes" autoplay="ap=1">http://example.com/videos/1.flv</video:player_loc>', $fileContent);
        $this->assertContains('<video:player_loc allow_embed="yes" autoplay="ap=1">http://example.com/videos/2.flv</video:player_loc>', $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteExtraContent()
    {
        $siteMapFile = $this->createSiteMapFile();

        $siteMapFile->writeUrl(
            'http://example.com/some',
            [],
            '<!-- extra-content -->'
        );

        $siteMapFile->close();

        $fileContent = file_get_contents($siteMapFile->getFullFileName());

        $this->assertContains('<!-- extra-content --></url>', $fileContent);
    }

    /**
     * @depends testWriteUrl
     */
    public function testWriteInMemory()
    {
        $siteMapFile = $this->createSiteMapFile();

        $siteMapFile->fileName = 'php://memory';

        $siteMapFile->writeUrl('http://example.com/foo');

        $fileContent = $siteMapFile->getContent();

        $this->assertContains('<?xml', $fileContent);
        $this->assertContains('http://example.com/foo', $fileContent);
        $this->assertContains('<urlset ', $fileContent);
        $this->assertContains('</urlset>', $fileContent);
    }
}


tests/TestCase.php
<?php

namespace yii2tech\tests\unit\sitemap;

use yii\helpers\ArrayHelper;
use Yii;
use yii\helpers\FileHelper;

/**
 * Base class for the test cases.
 */
class TestCase extends \PHPUnit\Framework\TestCase
{
    protected function setUp()
    {
        parent::setUp();
        $this->mockApplication();

        $testFilePath = $this->getTestFilePath();
        FileHelper::createDirectory($testFilePath);
    }

    protected function tearDown()
    {
        $testFilePath = $this->getTestFilePath();
        FileHelper::removeDirectory($testFilePath);

        $this->destroyApplication();
    }

    /**
     * Populates Yii::$app with a new application
     * The application will be destroyed on tearDown() automatically.
     * @param array $config The application configuration, if needed
     * @param string $appClass name of the application class to create
     */
    protected function mockApplication($config = [], $appClass = '\yii\console\Application')
    {
        new $appClass(ArrayHelper::merge([
            'id' => 'testapp',
            'basePath' => __DIR__,
            'vendorPath' => $this->getVendorPath(),
            'components' => [
                'urlManager' => [
                    'hostInfo' => 'http://test.com',
                    'baseUrl' => '/',
                    'scriptUrl' => '/index.php',
                ],
            ],
        ], $config));
    }

    /**
     * @return string vendor path
     */
    protected function getVendorPath()
    {
        return dirname(__DIR__) . '/vendor';
    }

    /**
     * Destroys application in Yii::$app by setting it to null.
     */
    protected function destroyApplication()
    {
        Yii::$app = null;
    }

    /**
     * Returns the test file path.
     * @return string file path.
     */
    protected function getTestFilePath()
    {
        return Yii::getAlias('@yii2tech/tests/unit/sitemap/runtime') . DIRECTORY_SEPARATOR . getmypid();
    }
}


tests/bootstrap.php
<?php

// ensure we get report on all possible php errors
error_reporting(-1);

define('YII_ENABLE_ERROR_HANDLER', false);
define('YII_DEBUG', true);
$_SERVER['SCRIPT_NAME'] = '/' . __DIR__;
$_SERVER['SCRIPT_FILENAME'] = __FILE__;

require_once(__DIR__ . '/../vendor/autoload.php');
require_once(__DIR__ . '/../vendor/yiisoft/yii2/Yii.php');

Yii::setAlias('@yii2tech/tests/unit/sitemap', __DIR__);
Yii::setAlias('@yii2tech/sitemap', dirname(__DIR__) . '/src');

require_once(__DIR__ . '/compatibility.php');

src/File.php
<?php
/**
 * @link https://github.com/yii2tech
 * @copyright Copyright (c) 2015 Yii2tech
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yii2tech\sitemap;

use yii\base\InvalidArgumentException;

/**
 * File is a helper to create the site map XML files.
 * Example:
 *
 * ```php
 * use yii2tech\sitemap\File;
 *
 * $siteMapFile = new File();
 * $siteMapFile->writeUrl(['site/index']);
 * $siteMapFile->writeUrl(['site/contact'], ['priority' => '0.4']);
 * $siteMapFile->writeUrl('http://mydomain.com/mycontroller/myaction', [
 *     'lastModified' => '2012-06-28',
 *     'changeFrequency' => 'daily',
 *     'priority' => '0.7'
 * ]);
 * ...
 * $siteMapFile->close();
 * ```
 *
 * @see BaseFile
 * @see http://www.sitemaps.org/
 *
 * @author Paul Klimov <klimov.paul@gmail.com>
 * @since 1.0
 */
class File extends BaseFile
{
    // Check frequency constants:
    const CHECK_FREQUENCY_ALWAYS = 'always';
    const CHECK_FREQUENCY_HOURLY = 'hourly';
    const CHECK_FREQUENCY_DAILY = 'daily';
    const CHECK_FREQUENCY_WEEKLY = 'weekly';
    const CHECK_FREQUENCY_MONTHLY = 'monthly';
    const CHECK_FREQUENCY_YEARLY = 'yearly';
    const CHECK_FREQUENCY_NEVER = 'never';

    /**
     * {@inheritdoc}
     */
    public $rootTag = [
        'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9',
    ];

    /**
     * @var array default options for {@see writeUrl()}.
     */
    public $defaultOptions = [];


    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        if (!empty($this->rootTag) && !isset($this->rootTag['tag'])) {
            $this->rootTag['tag'] = 'urlset';
        }
    }

    /**
     * Writes the URL block into the file.
     * @param string|array $url page URL or params.
     * @param array $options options list, valid options are:
     *
     * - 'lastModified' - string|int, last modified date in format Y-m-d or timestamp.
     * - 'changeFrequency' - string, page change frequency, the following values can be passed:
     *
     *   * always
     *   * hourly
     *   * daily
     *   * weekly
     *   * monthly
     *   * yearly
     *   * never
     *
     *   You may use constants defined in this class here.
     * - 'priority' - string|float URL search priority in range 0..1
     * - 'images' - array list of images bound to the URL, {@see composeImage()} for details.
     * - 'videos' - array list of videos bound to the URL, {@see composeVideo()} for details.
     *
     * @param string|null $extraContent extra XML content to be placed inside 'url' tag.
     * @return int the number of bytes written.
     */
    public function writeUrl($url, array $options = [], $extraContent = null)
    {
        $this->incrementEntriesCount();

        if (!is_string($url)) {
            $url = $this->getUrlManager()->createAbsoluteUrl($url);
        }

        $xmlCode = '<url>';
        $xmlCode .= "<loc>{$url}</loc>";

        if (($unrecognizedOptions = array_diff(array_keys($options), ['lastModified', 'changeFrequency', 'priority', 'images', 'videos'])) !== []) {
            throw new InvalidArgumentException('Unrecognized options: ' . implode(', ', $unrecognizedOptions));
        }

        $options = array_merge($this->defaultOptions, $options);

        if (isset($options['lastModified']) && ctype_digit($options['lastModified'])) {
            $options['lastModified'] = date('Y-m-d', $options['lastModified']);
        }

        if (isset($options['lastModified'])) {
            $xmlCode .= '<lastmod>' . $this->normalizeDateValue($options['lastModified']) . '</lastmod>';
        }
        if (isset($options['changeFrequency'])) {
            $xmlCode .= '<changefreq>' . $options['changeFrequency'] . '</changefreq>';
        }
        if (isset($options['priority'])) {
            $xmlCode .= '<priority>' . $options['priority'] . '</priority>';
        }

        if (!empty($options['images'])) {
            foreach ($options['images'] as $image) {
                $xmlCode .= $this->lineBreak . $this->composeImage($image);
            }
        }

        if (!empty($options['videos'])) {
            foreach ($options['videos'] as $video) {
                $xmlCode .= $this->lineBreak . $this->composeVideo($video);
            }
        }

        if ($extraContent !== null) {
            $xmlCode .= $extraContent;
        }

        $xmlCode .= '</url>';

        return $this->write($xmlCode);
    }

    /**
     * Creates XML code for image tag.
     * @see https://www.google.com/schemas/sitemap-image/1.1/
     *
     * @param array $image image options, valid options are:
     *
     * - 'url' - string
     * - 'title' - string
     * - 'caption' - string
     * - 'geoLocation' - string
     * - 'license' - string
     *
     * @return string XML code.
     * @since 1.1.0
     */
    protected function composeImage(array $image)
    {
        $xmlCode = '<image:image>';

        $xmlCode .= '<image:loc>' . $image['url'] . '</image:loc>';

        if (isset($image['title'])) {
            $xmlCode .= '<image:title>' . $image['title'] . '</image:title>';
        }
        if (isset($image['caption'])) {
            $xmlCode .= '<image:caption>' . $image['caption'] . '</image:caption>';
        }
        if (isset($image['geoLocation'])) {
            $xmlCode .= '<image:geo_location>' . $image['geoLocation'] . '</image:geo_location>';
        }
        if (isset($image['license'])) {
            $xmlCode .= '<image:license>' . $image['license'] . '</image:license>';
        }

        $xmlCode .= '</image:image>';

        return $xmlCode;
    }

    /**
     * Creates XML code for video tag.
     * @see https://www.google.com/schemas/sitemap-video/1.1/
     *
     * @param array $video video options, valid options are:
     *
     * - 'thumbnailUrl' - string, URL to the thumbnail
     * - 'title' - string, video page title
     * - 'description' - string, video page meta description
     * - 'contentUrl' - string
     * - 'duration' - int|string, video length in seconds
     * - 'expirationDate' - string|int
     * - 'rating' - string
     * - 'viewCount' - string|int
     * - 'publicationDate' - string|int
     * - 'familyFriendly' - string
     * - 'requiresSubscription' - string
     * - 'live' - string
     * - 'player' - array, options:
     *
     *   * 'url' - string, URL to raw video clip
     *   * 'allowEmbed' - bool|string
     *   * 'autoplay' - bool|string
     *
     * - 'restriction' - array, options:
     *
     *   * 'relationship' - string
     *   * 'restriction' - string
     *
     * - 'gallery' - array, options:
     *
     *   * 'title' - string
     *   * 'url' - string
     *
     * - 'price' - array, options:
     *
     *   * 'currency' - string
     *   * 'price' - string|float
     *
     * - 'uploader' - array, options:
     *
     *   * 'info' - string
     *   * 'uploader' - string
     *
     * @return string XML code.
     * @since 1.1.0
     */
    protected function composeVideo(array $video)
    {
        $xmlCode = '<video:video>';

        if (isset($video['thumbnailUrl'])) {
            $xmlCode .= '<video:thumbnail_loc>' . $video['thumbnailUrl'] . '</video:thumbnail_loc>';
        }
        if (isset($video['title'])) {
            $xmlCode .= '<video:title><![CDATA[' . $video['title'] . ']]></video:title>';
        }
        if (isset($video['description'])) {
            $xmlCode .= '<video:description><![CDATA[' . $video['description'] . ']]></video:description>';
        }
        if (isset($video['contentUrl'])) {
            $xmlCode .= '<video:content_loc>' . $video['contentUrl'] . '</video:content_loc>';
        }
        if (isset($video['duration'])) {
            $xmlCode .= '<video:duration>' . $video['duration'] . '</video:duration>';
        }
        if (isset($video['expirationDate'])) {
            $xmlCode .= '<video:expiration_date>' . $this->normalizeDateValue($video['expirationDate']) . '</video:expiration_date>';
        }
        if (isset($video['rating'])) {
            $xmlCode .= '<video:rating>' . $video['rating'] . '</video:rating>';
        }
        if (isset($video['viewCount'])) {
            $xmlCode .= '<video:view_count>' . $video['viewCount'] . '</video:view_count>';
        }
        if (isset($video['publicationDate'])) {
            $xmlCode .= '<video:publication_date>' . $this->normalizeDateValue($video['publicationDate']) . '</video:publication_date>';
        }
        if (isset($video['familyFriendly'])) {
            $xmlCode .= '<video:family_friendly>' . $video['familyFriendly'] . '</video:family_friendly>';
        }
        if (isset($video['requiresSubscription'])) {
            $xmlCode .= '<video:requires_subscription>' . $video['requiresSubscription'] . '</video:requires_subscription>';
        }
        if (isset($video['live'])) {
            $xmlCode .= '<video:live>' . $video['live'] . '</video:live>';
        }
        if (isset($video['player'])) {
            $xmlCode .= '<video:player_loc allow_embed="' . $this->normalizeBooleanValue($video['player']['allowEmbed']) . '" autoplay="' . $this->normalizeBooleanValue($video['player']['autoplay']) . '">'
                . $video['player']['url']
                . '</video:player_loc>';
        }
        if (isset($video['restriction'])) {
            $xmlCode .= '<video:restriction relationship="' . $video['restriction']['relationship'] . '">' . $video['restriction']['restriction'] . '</video:restriction>';
        }
        if (isset($video['gallery'])) {
            $xmlCode .= '<video:gallery_loc title="' . $video['gallery']['title'] . '">' . $video['gallery']['url'] . '</video:gallery_loc>';
        }
        if (isset($video['price'])) {
            $xmlCode .= '<video:price currency="' . $video['price']['currency'] . '">' . $video['price']['price'] . '</video:price>';
        }
        if (isset($video['uploader'])) {
            $xmlCode .= '<video:uploader info="' . $video['uploader']['info'] . '">' . $video['uploader']['uploader'] . '</video:uploader>';
        }

        $xmlCode .= '</video:video>';

        return $xmlCode;
    }
}
